src/Lykke.ClientGenerator/Infrastructure/AopProxy.cs

```diff
@@ -0,0 +1,187 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+
+namespace Lykke.ClientGenerator.Infrastructure
+{
+    /// <summary>
+    /// Helper to generate a calls wrapping proxy
+    /// </summary>
+    public class AopProxy : DispatchProxy
+    {
+        private static readonly ConcurrentDictionary<MethodInfo, MethodHelperFuncs> _cache =
+            new ConcurrentDictionary<MethodInfo, MethodHelperFuncs>();
+
+        private MethodCallHandler[] _handlers;
+        private object _decorated;
+
+        /// <summary>
+        /// Describes the delegate handling the method call 
+        /// </summary>
+        public delegate Task<object> MethodCallHandler(MethodInfo targetMethod, object[] args,
+            Func<Task<object>> innerHandler);
+
+        private void Initialize(object decorated, MethodCallHandler[] handlers)
+        {
+            _decorated = decorated;
+            _handlers = handlers;
+        }
+
+        /// <inheritdoc />
+        protected override object Invoke(MethodInfo targetMethod, object[] args)
+        {
+            // ReSharper disable once GenericEnumeratorNotDisposed
+            var enumerator = _handlers.AsEnumerable().GetEnumerator();
+            enumerator.MoveNext();
```

Comment:
what if _handlers contains no elements?

Reply:
Well, it's checked on the caller level, but it's better to put a guard on the type initialization.

Reply:
It turned out this guard already exists